domain/modelagent/state/modelstate.go

```diff
@@ -137,35 +194,126 @@ func (st *State) GetMachineUUID(ctx context.Context, name machine.Name) (string,
 	return uuid.UUID, nil
 }
 
-// GetMachineTargetAgentVersion returns the target agent version for the specified machine.
+// GetMachineRunningAgentBinaryVersion reports the currently set agent binary
+// version value for a machine. The following errors can be expected:
+// - [machineerrors.MachineNotFound] when the machine being asked for does
+// not exist.
+// - [modelagenterrors.AgentVersionNotFound] when no
+// running agent version has been set for the given machine.
+func (st *State) GetMachineRunningAgentBinaryVersion(
+	ctx context.Context,
+	uuid string,
+) (coreagentbinary.Version, error) {
+	db, err := st.DB()
+	if err != nil {
+		return coreagentbinary.Version{}, errors.Capture(err)
+	}
+
+	rval := machineAgentVersionInfo{}
+	machineUUID := machineUUIDRef{
+		UUID: uuid,
+	}
+
+	stmt, err := st.Prepare(`
+SELECT &machineAgentVersionInfo.*
+FROM v_machine_agent_version
+WHERE machine_uuid = $machineUUIDRef.machine_uuid
+`, rval, machineUUID)
+	if err != nil {
+		return coreagentbinary.Version{}, errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := st.checkMachineNotDead(ctx, tx, uuid)
+		if errors.Is(err, machineerrors.MachineNotFound) {
```

Comment:
Did you want to return early if the machine is dead?

Reply:
I am inclined not to concern ourselves with life here. This is a value read and as we don't understand the context for which the client is asking the question it seems better to report back what we had. Migration won't happen for dying machines.

I am going to do Harry's feedback above and break the duality of this function apart. I got a bit lazy in the end.